### DISCUSSION TITLE
Problem in evaluating flow rate from stokes flow simulation

### DISCUSSION THREAD

**Discussion by BrinthanK** [PMEAL/OpenPNM#2930] [Q&A] [answered]
We have a positive pressure gradient from inlet to outlet across a pore networks. Therefore, we expect the fluid to flow **in** but we get a value of **Q>0**, which according to OpenPNM means that the fluid is **flowing out**. Is there any reason as to why this is the case? 

For context we are evaluating the permeability coefficient using the Darcy's law for which the flow rate is required. We are attaching the code snippit and the outcome of the code. 

Thank you in advance. 
![01](https://github.com/user-attachments/assets/6da9c5c3-fe57-482f-ad70-28f9a8839ffd)
![02](https://github.com/user-attachments/assets/e3f8308e-c00e-414f-87c8-748d6a32e428)

**Comment by jgostick** [comment, marked as answer]
You have a lot of extra negatives signs in your code snippet above.  I think they look right, but possibly they are the problem.  The other problem might be where you are computing the rate.   If you use the inlet pores the rate should be positive, but if you use the outlet pore it will be negative (with a magnitude equal to the inlet).

**Comment by BrinthanK** [comment]
Hey Jeff, thank you for your response. I have an additional question on top of this, that is, for the stokes law calculation what boundary values (i.e., flow.set_value_BC) should be, is it pressure or velocity?  

Thank you.

**Reply by ma-sadeghi** [reply]
It is pressure. If you want to set the flow rate, you can use `set_rate_BC`.

**Reply by spaul98** [reply]
Hi Sadeghi,
If I use the set_rate_BC on one boundary and P_out for the other and want to calculate the pressure drop from Stokes flow, is that possible? I see the flow.rate() option, but I was wondering if there is a function to calculate the delP. Please let me know.
Thank you.

**Reply by jgostick** [reply]
It is kinda possible.  If you put a flow rate on one side, the the pressure is free to become whatever it wants, which means that each pore will have a slightly different pressure.  You'll have to take the average.  If you set pressure on both sides, then you'll know the pressure with certainty, and you can find the flow, which is the sum of flow in all inlet pores.  This is assured to be correct, compared to finding an average pressure which will only be an approximation.

**Reply by spaul98** [reply]
I understand. Thank you for the response. The problem is with the pressure BCs, I am getting a negative flow rate at the inlet in x-direction even when my pressure gradient is positive. I have checked the BCs and I still do not know why that is so.

**Reply by jgostick** [reply]
You should get a positive rate at the inlets and a negative rate at the outlets, and they should have equal magnitude.  So `sf.rate(pores=pn['pore.left']) == `sf.rate(pores=pn['pore.right'])`, assuming `'pore.left'` were used as inlets and `'pore.right'` were used as outlets.

**Reply by jgostick** [reply]
When in doubt you can always do a sample simulation on a simple network, like in one of the examples: https://openpnm.org/examples/tutorials/08_simulating_transport.html

**Reply by jgostick** [reply]
It just occurred to me that your problem might be caused by negative conductances.  This can happen if your pores are too large and they overlap each other, resulting in a throat length that is negative.  You should check this with something like `np.sum(sf['throat.hydraulic_conductance'] < 0])`.  If it's not zero, then that is a problem...hopefully your only problem.

**Reply by spaul98** [reply]
For some reason, I get a 'key error' when I try to run the above line of code; (image attached)
![SS_key_error](https://github.com/user-attachments/assets/9428aad0-1fa1-44cf-aefb-9ad1b054e4d6)

**Reply by spaul98** [reply]
I am trying to simulate with a rate BC and P_out BC, to evaluate delP, as I mentioned before. Can you please confirm if the code below is atleast logically correct?
![SS_delP](https://github.com/user-attachments/assets/43583367-23e8-4dca-96cc-dc6bdc8391a2)

**Reply by spaul98** [reply]
Hi, 
I just wanted to let you know that I think the top equation on this link: https://openpnm.org/examples/applications/effective_diffusivity_and_tortuosity.html is not represented correctly. You have D_ab on the left hand side of the equation but I think it should be  on the right hand side and D_eff should be on the left hand side. Further down you have it correct. The description is also alright. 
Thank you.

**Reply by jgostick** [reply]
good catch, thanks! I already fixed it.  Nice to have such an easy fix.
